### PR TITLE
add tests for chessboard corner detection

### PR DESCRIPTION
**Pull request by yakir12**
This is my first PR in hopes to add a full test-suit for the opencv camera calibration functions. My secret motive is that I've found that `OpenCV.calibrateCamera` randomly fails (i.e. return nonsensical values), and I hope to unearth the issue together with the maintainers of this repo.

### CODE REVIEW

**Comment by yakir12** [comment]
Super weird. I noticed that just `chess5.png` was failing, and that it was first in the list... So I tried running the detection once without comparing its results to the saved coordinates of the corners, and now it works. It's as if there is a need for some initialization for it to work properly...?

**Comment by yakir12** [comment]
(@timholy, you asked me to ping you directly)

File: test/test_corner_detection.jl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        cols = parse(Int, readuntil(o, "\n"))
+        readuntil(o, "[")
+        corners = [Float32[] for i in 1:cols, j in 1:rows]
+        for i in 1:cols*rows, j in 1:2
+            if i == cols*rows && j == 2
+                break
+            end
+            push!(corners[i], parse(Float32, readuntil(o, ",")))
+        end
+        push!(corners[cols*rows], parse(Float32, readuntil(o, "]")))
```
Comment (timholy) [inline test/test_corner_detection.jl:26]: If I understand correctly, this parses a matrix like [this](https://github.com/opencv/opencv_extra/blob/a74cf6bae7fd75d91282b877c559168b3a62148a/testdata/cv/cameracalibration/chess_corners1.dat#L8-L25) into a matrix-of-vectors where each vector holds a single value. I'm not familiar with OpenCV, so can you explain the reason for this choice? Why not `readdlm` the matrix directly?

File: test/test_corner_detection.jl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+function calc_error(ps1, ps2)
+    s = 0.0
+    for (p1, p2) in zip(ps1, ps2)
+        s += LinearAlgebra.norm_sqr(p1 .- p2)
+    end
+    sqrt(s/length(ps1))
+end
+
+function test_one(img_file, data_file)
```
Comment (timholy) [inline test/test_corner_detection.jl:51]: Since this doesn't run a test, I'd call this `calc_error(img_file::AbstractString, data_file::AbstractString)`.

File: test/test_corner_detection.jl [hunk 3]
```diff
@@ -0,0 +1,73 @@
+function _detect_corners(file, n_corners)
+    img = load(file)
+    gry = img[1:1, :, :]
+    cv_n_corners = OpenCV.Size{Int32}(n_corners...)
+    _cv_corners = OpenCV.Mat(Array{Float32}(undef, 2, 1, prod(n_corners)))
```
Comment (timholy) [inline test/test_corner_detection.jl:5]: Out of curiosity, could this be declared `OpenCV.Mat(Matrix{Tuple{Float32,Float32}}(undef, n_corners))`? And then avoid the reshaping/`eachslice` below?

File: test/test_corner_detection.jl [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        end
+        push!(corners[cols*rows], parse(Float32, readuntil(o, "]")))
+
+        return reverse(permutedims(corners, (2, 1)); dims = 1), (rows, cols)
+    end
+end
+
+function get_list(file)
+    doc = read(file, LazyNode)
+    str = filter(≠('\"'), simple_value(doc[end][end]))
```
Comment (timholy) [inline test/test_corner_detection.jl:34]:
```suggestion
    str = filter(≠('"'), simple_value(doc[end][end]))
```

**Review by timholy — COMMENTED** [review]
Thanks for the PR! It mostly looks good. I've left a few code comments, but one overall comment:

> together with the maintainers of this repo

This repo has no maintainers. There are people like me who have merge privileges via JuliaImages, but I don't know or use OpenCV. It would be great if this package got a real maintainer.

I don't know how to fix the "must be initialized" bug, sadly.

File: test/test_corner_detection.jl (hunk 1 above)
Comment (timholy) [reply test/test_corner_detection.jl:26]: I think you can use `",]"` as the delimiter for `readuntil` and avoid the need for special-casing `i==cols*rows`.

File: test/test_corner_detection.jl (hunk 3 above)
Comment (yakir12) [reply test/test_corner_detection.jl:5]: Currently, `OpenCV.Mat` can only accept 3D arrays, not matrices:
```julia
  MethodError: no method matching OpenCV.Mat(::Matrix{Tuple{Float32, Float32}})
  The type `OpenCV.Mat` exists, but no method is defined for this combination of argument types when trying to construct it.
  
  Closest candidates are:
    OpenCV.Mat(::AbstractArray{T, 3}) where T<:Union{Float32, Float64, Int16, Int32, Int8, UInt16, UInt8}
     @ OpenCV ~/.julia/artifacts/dcc70861ed0ffc5427898f6aeb436620f33fa02f/OpenCV/src/Mat.jl:13
```
So that's not an option. But even if it was, the values in the input arrays are never mutated in-place after the function call, so I still need to deal with the output arrays, and their shape and form is as it is.

File: test/test_corner_detection.jl (hunk 2 above)
Comment (yakir12) [reply test/test_corner_detection.jl:51]: I totally agree.

File: test/test_corner_detection.jl (hunk 1 above)
Comment (yakir12) [reply test/test_corner_detection.jl]: I had the same thought, it's simpler now (my only reason to do it this way was that it takes less memory, but hardly worth the effort). The real solution would have been if https://github.com/JuliaData/YAML.jl would have parsed that file, but it doesn't:
```julia
julia> using YAML

julia> data = YAML.load_file("chess_corners1.dat")
ERROR: expected '<document start>' but found YAML.DirectiveToken at nothing
Stacktrace:
  [1] parse_document_start(stream::YAML.EventStream)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/parser.jl:176
  [2] parse_implicit_document_start(stream::YAML.EventStream)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/parser.jl:153
  [3] peek(stream::YAML.EventStream)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/parser.jl:52
  [4] compose_document(composer::YAML.Composer)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/composer.jl:44
  [5] compose(events::YAML.EventStream, resolver::YAML.Resolver)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/composer.jl:33
  [6] load(tokenstream::YAML.TokenStream, constructor::YAML.Constructor)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/YAML.jl:77
  [7] load
    @ ~/.julia/packages/YAML/XZZ1Y/src/YAML.jl:84 [inlined]
  [8] load(input::IOStream, more_constructors::Nothing, multi_constructors::Dict{Any, Any}; kwargs::@Kwargs{})
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/YAML.jl:87
  [9] load (repeats 3 times)
    @ ~/.julia/packages/YAML/XZZ1Y/src/YAML.jl:87 [inlined]
 [10] #16
    @ ~/.julia/packages/YAML/XZZ1Y/src/YAML.jl:155 [inlined]
 [11] open(::YAML.var"#16#17"{@Kwargs{}, Tuple{}}, ::String, ::Vararg{String}; kwargs::@Kwargs{})
    @ Base ./io.jl:410
 [12] open
    @ ./io.jl:407 [inlined]
 [13] load_file(::String)
    @ YAML ~/.julia/packages/YAML/XZZ1Y/src/YAML.jl:153
 [14] top-level scope
    @ REPL[4]:1
```

File: test/test_corner_detection.jl [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    calc_error(corners, detected_corners)
+end
+
+@testset "detecting corners" begin
+
+    path = joinpath(test_dir, "cameracalibration")
+    list = get_list(joinpath(path, "chessboard_list.dat"))
+
+    k, v = first(list)
+    calc_error(joinpath(path, k), joinpath(path, v)) # why do we need this?
```
Comment (yakir12) [inline test/test_corner_detection.jl:59]: This is the main issue. why do we need to run this first...?

**Comment by yakir12** [comment]
I agree and understand. I'm not even sure there's any need to know much about OpenCV to fix this. As I've noticed elsewhere, `opencv` via `PythonCall` works just fine, so it seems like something in the porting of opencv to Julia went a little off. My goal is to add a bunch of tests to this module that might highlight and insulate the issues in hand so that it'll be easier for anyone to resolve the issue. The main one being this one: https://github.com/JuliaImages/OpenCV.jl/pull/52/files#r1899141463

File: test/test_corner_detection.jl (hunk 3 above)
Comment (timholy) [reply test/test_corner_detection.jl:7]: You should be able to do
```julia
corners = Matrix{Tuple{Float32,Float32}}(undef, n_corners)
ret, cv_corners = OpenCV.findChessboardCorners(gry, cv_n_corners, OpenCV.Mat(reinterpret(reshape, Float32, corners)), 0)
@assert ret "Failed to detect any corners!"
return corners
```
if you think that's a little bit clearer/cleaner.

Comment (yakir12) [reply test/test_corner_detection.jl:7]: It is clearer, does not error (and TIL about `reinterpret(reshape...`), and I would love to use it, but the call to `findChessboardCorners` does not mutate its arguments:
```julia
corners = Matrix{Tuple{Float32,Float32}}(undef, n_corners)
before = deepcopy(corners)
ret, cv_corners = OpenCV.findChessboardCorners(gry, cv_n_corners, OpenCV.Mat(reinterpret(reshape, Float32, corners)), 0)

julia> all(corners .=== before)
true
```
So we would anyways need to reshape and slice the return `cv_corners` value, like this:
```julia
function detect_corners(file, n_corners)
    img = load(file)
    gry = img[1:1, :, :]
    cv_n_corners = OpenCV.Size{Int32}(n_corners...)
    _corners = Matrix{Tuple{Float32,Float32}}(undef, n_corners)
    ret, cv_corners = OpenCV.findChessboardCorners(gry, cv_n_corners, OpenCV.Mat(reinterpret(reshape, Float32, _corners)), 0)
    corners = reshape(Point.(eachslice(cv_corners, dims = 3)), n_corners)
    return corners
end
```

I've noticed this issue before (cause I tried something similar to what you suggested). This is surly another bug..?

Comment (timholy) [reply test/test_corner_detection.jl:7]: Must be. Otherwise why pass an array pre-allocated for the output?

Comment (yakir12) [reply test/test_corner_detection.jl:7]: I understand why you'd think so, I did too, but lo and behold:
```julia
julia> hash(_corners)
0x88d97a55d4ebfe14

julia> before = deepcopy(_corners)
7×5 Matrix{Tuple{Float32, Float32}}:
 (1.0f-45, 0.0)  (0.0, 0.0)      (1.0f-45, 0.0)  (0.0, 0.0)      (0.0, 0.0)
 (3.0f-45, 0.0)  (6.0f-45, 0.0)  (0.0, 0.0)      (1.0f-45, 0.0)  (0.0, 0.0)
 (0.0, 0.0)      (1.0f-45, 0.0)  (1.0f-45, 0.0)  (0.0, 0.0)      (0.0, 0.0)
 (1.0f-45, 0.0)  (0.0, 0.0)      (1.0f-45, 0.0)  (1.0f-45, 0.0)  (1.0f-45, 0.0)
 (1.0f-45, 0.0)  (1.0f-45, 0.0)  (1.0f-45, 0.0)  (0.0, 0.0)      (1.0f-45, 0.0)
 (4.0f-45, 0.0)  (0.0, 0.0)      (1.0f-45, 0.0)  (3.0f-45, 0.0)  (0.0, 0.0)
 (1.0f-45, 0.0)  (0.0, 0.0)      (0.0, 0.0)      (1.0f-45, 0.0)  (0.0, 0.0)

julia> ret, cv_corners = OpenCV.findChessboardCorners(gry, cv_n_corners, OpenCV.Mat(reinterpret(reshape, Float32, _corners)), 0)
(true, Float32[89.02004; 65.43221;;; 101.30311; 66.39866;;; 114.23407; 67.06309;;; … ;;; 153.37643; 118.27199;;; 166.67719; 119.35007;;; 180.67068; 120.32631])

julia> hash(_corners)
0x88d97a55d4ebfe14

julia> all(_corners .=== before)
true
```

Comment (timholy) [reply test/test_corner_detection.jl:7]: So what is that argument for?

Comment (yakir12) [reply test/test_corner_detection.jl:7]: I don't know and I don't know how to find out, everything is wrapped [here](https://github.com/opencv/opencv_contrib/tree/4.x/modules/julia). I suppose we could turn a blind eye and just follow their [example](https://github.com/opencv/opencv_contrib/blob/4.x/modules/julia/samples/chessboard_corners.jl):
```julia
ret, corners = cv.findChessboardCorners(img, cv.Size{Int32}(7,5))
```

Comment (timholy) [reply test/test_corner_detection.jl:7]: At least the 2-arg form doesn't pretend to be passing in the output array, so I'd go with that if you can't figure out the source of the problem. But from the [OpenCV C++ docs for `findChessboardCorners`](https://docs.opencv.org/4.x/d9/d0c/group__calib3d.html#ga93efa9b0aa890de240ca32b11253dd4a) it clearly is supposed to be the output array.

Comment (yakir12) [reply test/test_corner_detection.jl]: (Please wait before merging, I want to try something new)

Comment (yakir12) [reply test/test_corner_detection.jl]: OK! Fixed it. The issue was that when using `reinterpret(reshape...` the shape of the array, (2, 35), wasn't identical to what `findChessboardCorners` was expecting, (2, 1, 35). 
I also removed the weird initialization step, and it worked. This is promising, maybe I'll be able to fix the rest of the camera calibration functions.

Please let me know if you have any additional suggestions, otherwise I'm pleased with this version.

**Comment by timholy** [comment]
Thanks for the contribution!
